src/main/java/fi/helsinki/cs/tmc/core/communication/ExerciseDownloader.java

```diff
@@ -85,19 +98,38 @@ public ExerciseDownloader(UrlCommunicator urlCommunicator, TmcApi tmcApi) {
      * @param exercises list of exercises which will be downloaded, list is parsed from json.
      * @param path server path to exercises.
      * @param folderName folder name of where exercises will be extracted (for example course name)
+     * @return
      */
     public Optional<List<Exercise>> downloadFiles(
             List<Exercise> exercises, String path, String folderName) {
-        List<Exercise> downloadedExercises = new ArrayList<>();
-        path = createCourseFolder(path, folderName);
-        for (Exercise exercise : exercises) {
-            boolean success = handleSingleExercise(exercise, path);
-            if (success) {
-                downloadedExercises.add(exercise);
-            }
+        final List<Future<Optional<Exercise>>> futures = new ArrayList<>();
+        final String coursePath = createCourseFolder(path, folderName);
+        for (final Exercise exercise : exercises) {
+            Callable<Optional<Exercise>> downloadHandler
+                    = new SingleExerciseDownloadHandler(exercise, coursePath);
+            futures.add(downloadThreadPool.submit(downloadHandler));
         }
+        return collectExerciseFutures(futures);
+    }
 
-        return Optional.of(downloadedExercises);
+    private Optional<List<Exercise>> collectExerciseFutures(
+            List<Future<Optional<Exercise>>> futures) {
+        List<Exercise> exercises = new ArrayList<>();
+        for (Future<Optional<Exercise>> future : futures) {
+            try {
+                Optional<Exercise> exercise = future.get();
+                if (exercise.isPresent()) {
+                    exercises.add(exercise.get());
+                }
+            }
+            catch (ExecutionException ex) {
+                log.error("Failed to handle exercise: {}", ex);
+            }
+            catch (InterruptedException ex) {
+                log.error("Handling of exercise was interrupted: {}", ex);
+            }
+        }
+        return Optionals.ofNonEmpty(exercises);
```

Comment:
What would be the best way to do this? Just throw an exception here and handle exception in the caller?